sound/soc/sof/ipc4-topology.c

```diff
@@ -1483,8 +1508,9 @@ static int sof_ipc4_init_input_audio_fmt(struct snd_sof_dev *sdev,
 		rate = fmt->sampling_frequency;
 		channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(fmt->fmt_cfg);
 		valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+		type = SOF_IPC4_AUDIO_FORMAT_CFG_SAMPLE_TYPE(fmt->fmt_cfg);
 		if (params_rate(params) == rate && params_channels(params) == channels &&
-		    sample_valid_bits == valid_bits)
+		    sample_valid_bits == valid_bits && sample_type == type)
```

Comment:
@singalsu do you have a corresponsing topology change for this? Dont we default to LSB_INTEGER in topology?

Reply:
Yes, the topology that I use for testing is in https://github.com/thesofproject/sof/pull/9980 .

The MSB integer and LSB integer usage is confusing. I would use in topology signed integer for s16 and s32. At least that's what comes to my mind first. Though they would be also valid MSB integer and LSB integer since the valid size is the same as container type. The ALSA S24_4LE as low part aligned would be LSB integer.

I need this check to handle more special types to separate A-law from integer 8-bit (that is also special as unsigned) . Also float (we miss support)  is 32 bits valid and container size and need be set to IPC4 float type.

Reply:
@ranj063 The playback direction works now, the issue was in topology. Yes, the topologies default to LSB_INTEGER and I've not changed that. I decided to use IPC4_TYPE_UNSIGNED_INTEGER for ALSA U8 since there is also ALSA S8. If need support for that it would be (maybe bestas ) IPC4_TYPE_SIGNED_INTEGER. Since it's a byte IPC4_TYPE_LSB_INTEGER would be confusing. So far I'm not aware of any wav files with S8 format, so I'm not planning to add it now.

I currently work with capture direction. With the latest PR here, I can run capture with same formats but I haven't checked that it's correct yet.

Reply:
@singalsu I think we need to test this patch alone (as a separate PR) with the current topologies to make sure we pass isnt it? We've never included the sample_type in the tests before and we should make sure that it is correct in existing topologies before we start adding support for 8-bit formats

Reply:
Yep, in tests below seems nocodec fails ([  264.082122] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Seeking for type -949225600). I need to fix those uninitialized types first in topologies.

Reply:
Actually above was a bug in this patch with use of unset type. New try soon with this kernel patches CI.